{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FACTORBOOK motifs processing SOP for FILER integration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Introduction\n",
    "\n",
    "Factorbook is a transcription factor (TF)-centric repository of all ENCODE ChIP-seq datasets on TF-binding regions, as well as the rich analysis results of these data.\n",
    "\n",
    "#### 1.1 Factorbook motifs:\n",
    "Factorbook motifs were identified by applying MEME to the top 500 IDR thresholded ChIP-seq peaks from more than 3000 ENCODE ChIP-seq experiments. Five motifs were identified per experiment. These motifs were then filtered for quality using peak centrality and enrichment metrics. In total, there are 6,069 motifs available.\n",
    "\n",
    "[Factorbook motif catalog](https://downloads.wenglab.org/factorbook-download/complete-factorbook-catalog.meme.gz)\n",
    "\n",
    "[Associated metadata](https://storage.googleapis.com/gcp.wenglab.org/factorbook_chipseq_meme_motifs.tsv)\n",
    "\n",
    "#### 1.2. Factorbook motif annotation for FILER\n",
    "\n",
    "Factorbook motifs are downloaded and processed for FILER integration. This processing involves annotating the factorbook motifs by identifying overlaps with known motifs from the HOCOMOCO and JASPAR databases.\n",
    "\n",
    "TOMTOM was used to compare Factorbook MEME motifs against the HOCOMOCO and JASPAR catalogs. From tomtom results, the motifs with the lowest q-value were selected as the most confident overlaps between Factorbook motifs and the known motifs in HOCOMOCO and JASPAR. These selected motifs were then filtered for the condition that the query sequence is contained within the target sequence, or vice versa.\n",
    "\n",
    "The purpose of this SOP is to document the steps and processes involved in processing and annotating Facorbook motifs.\n",
    "\n",
    "#### 1.3 Data processing workflow\n",
    "\n",
    "The image below illustrates the entire workflow involved in processing and annotating factorbook motifs for FILER integration.\n",
    "\n",
    "![title](/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/fb_workflow.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## 2.Requirements\n",
    "\n",
    "### 2.1 ENCODE metadata:\n",
    "\n",
    "ENCODE metadata is necessary to construct the download URL for each of the factorbook motifs. For example, factorbook motif ID (e.g., ENCFF674XTY_CMTBCTGGGARTTGTAG) is generated by combining the ENCODE BED file ID with the consensus sequence. \n",
    "\n",
    "#### 2.1.1 metadata download:\n",
    "Select all the bed and bigBed tracks available in the ENCODE experiment matrix. Used the below selection criteria:\n",
    "\n",
    "- Biosample: Organism --> Homo sapiens\n",
    "- Quality: Status --> released, archived, revoked\n",
    "- Analysis: Available file types --> All bed and bigBed file types\n",
    "- Analysis: Genome Assembly --> hg19, GRCh38\n",
    "\n",
    "The following URL is pre-selected to meet the above criteria on the ENCODE data portal:\n",
    "\n",
    "[ENCODE metadata](https://www.encodeproject.org/search/?type=Experiment&status=released&perturbed=false&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens&perturbed=true&status=archived&status=revoked&files.file_type=bed+narrowPeak&files.file_type=bigBed+narrowPeak&files.file_type=bed+idr_ranked_peak&files.file_type=bed+broadPeak&files.file_type=bed+bed3%2B&files.file_type=bigBed+broadPeak&files.file_type=bigBed+bed3%2B&files.file_type=bed+bedRnaElements&files.file_type=bigBed+bedRnaElements&files.file_type=bed+tss_peak&files.file_type=bigBed+tss_peak&files.file_type=bed+bedMethyl&files.file_type=bigBed+bedMethyl&files.file_type=bed+bed9%2B&files.file_type=bed+bedGraph&files.file_type=bigInteract&files.file_type=bed+idr_peak&files.file_type=bigBed+idr_peak&files.file_type=bigBed+bed9%2B&files.file_type=bed+bed9&files.file_type=bigBed+bed9&files.file_type=bed+bedLogR&files.file_type=bigBed+bedLogR&files.file_type=bed+bed12&files.file_type=bigBed+bed12&files.file_type=bigBed+bed6%2B&files.file_type=bed+bed6%2B&files.file_type=bed+bedExonScore&files.file_type=bigBed+bedExonScore&files.file_type=bed+peptideMapping&files.file_type=bigBed+peptideMapping&files.file_type=bed+bed3&files.file_type=bigBed+bed3&files.file_type=bed+modPepMap&files.file_type=bed+pepMap&files.file_type=bigBed+modPepMap&files.file_type=bigBed+pepMap)\n",
    "\n",
    "### 2.2 Tomtom: Motif comparison tool\n",
    "\n",
    "Tomtom compares one or more motifs against a database of known motifs (e.g., JASPAR). \n",
    "\n",
    "[MEME suite V5.5.5](https://meme-suite.org/meme/doc/download.html)\n",
    "\n",
    "### 2.3 Samtools\n",
    "\n",
    "[Samtools V1.17](https://www.htslib.org/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. FILER data processing pipeline\n",
    "\n",
    "The image below provides an overview of the data processing pipeline involved in factorbook motifs processing for FILER integration, outlining the inputs, outputs, and how each step is interconnected.\n",
    "\n",
    "\n",
    "![title](/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/fb_pipeline.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "### 3.1 Collect ENCODE experiment IDs from Factorbook motif catalog metadata\n",
    "\n",
    "This script extracts ENCODE experiment IDs from the Factorbook motif metadata. These IDs are used in subsequent steps to retrieve and process the associated ChIP-seq data.\n",
    "\n",
    "Script: step1-get_ENCODE_experiments.sh\n",
    "\n",
    "Input: factorbook motif metadata - factorbook_chipseq_meme_motifs.tsv\n",
    "\n",
    "Output: ENCODE_experimentIDs.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Input Files ###\n",
      "\n",
      "factorbook_chipseq_meme_motifs.tsv:\n",
      "\n",
      "Animal\tBiosample\tTarget\tIdentifier\tConsensus\tSample Type\tInformation\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tAAAGAGGAASTGAAA\tcell line\t\"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tCCTCCTAGCCCTRACACACAGCTGGG\tcell line\t\"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tSCBSBCMSBBCCRVCCHSCSVKCCYKSVCC\tcell line\t\"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tGCSYGGGGCAAGTGACCGTGCGTGTAAAGG\tcell line\t\"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tTGGACTTTGRACYYW\tcell line\t\"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
      "\n",
      "### Output File ###\n",
      "\n",
      "ENCODE_experimentIDs.txt:\n",
      "\n",
      "ENCSR000AHD\n",
      "ENCSR000AHF\n",
      "ENCSR000AIB\n",
      "ENCSR000AIC\n",
      "ENCSR000AID\n",
      "ENCSR000AIE\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "def display_file_head(file_path):\n",
    "    try:\n",
    "        file_name = os.path.basename(file_path)  \n",
    "        print(f\"\\n{file_name}:\\n\")  \n",
    "        with open(file_path, 'r') as file:\n",
    "            for _ in range(6): \n",
    "                print(file.readline().strip())\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {file_path}\")\n",
    "    except Exception as e:\n",
    "        print(f\"Error reading file {file_path}: {e}\")\n",
    "\n",
    "factorbook_metadata = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/factorbook_chipseq_meme_motifs.tsv\"\n",
    "encode_exp = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/ENCODE_experimentIDs.txt\"\n",
    "\n",
    "print(\"### Input Files ###\")\n",
    "display_file_head(factorbook_metadata)\n",
    "\n",
    "print(\"\\n### Output File ###\")\n",
    "display_file_head(encode_exp)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Process ENCODE metadata to map IDR thresholded peak BED file IDs for all ChIP-seq experiments\n",
    "\n",
    "This script processes the ENCODE metadata and maps IDR thresholded peak BED file IDs to their corresponding ChIP-seq experiments, it then constructs Factorbook motif IDs by combining the ENCODE BED file ID with the consensus sequence for each experiment. \n",
    "\n",
    "Script: step2-bedtobigBed_mapping.py\n",
    "\n",
    "Input: \n",
    "1. ENCODE metadata from section 2.1\n",
    "2. factorbook motifs metadata from section 1.1\n",
    "3. ENCODE experiment IDs from section 3.1\n",
    "\n",
    "Output: bedtobigBed_mapping.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Input Files ###\n",
      "\n",
      "filtered_metadata.txt:\n",
      "\n",
      "File accession\tFile type\tFile format type\tOutput type\tExperiment accession\tBiological replicate(s)\tDerived from\n",
      "ENCFF105HVH\tbed\tnarrowPeak\tIDR thresholded peaks\tENCSR228ELU\t2\t/files/ENCFF604PVE/, /files/ENCFF872HPF/, /files/ENCFF356LFX/\n",
      "ENCFF214SNH\tbed\tnarrowPeak\tIDR thresholded peaks\tENCSR228ELU\t1, 2\t/files/ENCFF604PVE/, /files/ENCFF932FOK/, /files/ENCFF872HPF/, /files/ENCFF356LFX/\n",
      "ENCFF204NXZ\tbed\tidr_ranked_peak\tIDR ranked peaks\tENCSR228ELU\t1, 2\t/files/ENCFF932FOK/, /files/ENCFF872HPF/, /files/ENCFF604PVE/\n",
      "ENCFF254RJL\tbigBed\tnarrowPeak\tIDR thresholded peaks\tENCSR228ELU\t1, 2\t/files/ENCFF214SNH/\n",
      "ENCFF228MFI\tbed\tidr_ranked_peak\tIDR ranked peaks\tENCSR228ELU\t2\t/files/ENCFF872HPF/, /files/ENCFF604PVE/\n",
      "\n",
      "factorbook_chipseq_meme_motifs.tsv:\n",
      "\n",
      "Animal\tBiosample\tTarget\tIdentifier\tConsensus\tSample Type\tInformation\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tAAAGAGGAASTGAAA\tcell line\t\"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tCCTCCTAGCCCTRACACACAGCTGGG\tcell line\t\"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tSCBSBCMSBBCCRVCCHSCSVKCCYKSVCC\tcell line\t\"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tGCSYGGGGCAAGTGACCGTGCGTGTAAAGG\tcell line\t\"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tTGGACTTTGRACYYW\tcell line\t\"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
      "\n",
      "ENCODE_experimentIDs.txt:\n",
      "\n",
      "ENCSR000AHD\n",
      "ENCSR000AHF\n",
      "ENCSR000AIB\n",
      "ENCSR000AIC\n",
      "ENCSR000AID\n",
      "ENCSR000AIE\n",
      "\n",
      "### Output File ###\n",
      "\n",
      "bedtobigBed_mapping.txt:\n",
      "\n",
      "Animal\tBiosample\tTarget\tExperiment_ID\tExperiment_ID\tbigBed_Accession\tBED_Accession\tConsensus\tSample_Type\tBED_Consensus\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tAAAGAGGAASTGAAA\tcell line\tENCFF002XEC_AAAGAGGAASTGAAA\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tCCTCCTAGCCCTRACACACAGCTGGG\tcell line\tENCFF002XEC_CCTCCTAGCCCTRACACACAGCTGGG\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tSCBSBCMSBBCCRVCCHSCSVKCCYKSVCC\tcell line\tENCFF002XEC_SCBSBCMSBBCCRVCCHSCSVKCCYKSVCC\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tGCSYGGGGCAAGTGACCGTGCGTGTAAAGG\tcell line\tENCFF002XEC_GCSYGGGGCAAGTGACCGTGCGTGTAAAGG\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tTGGACTTTGRACYYW\tcell line\tENCFF002XEC_TGGACTTTGRACYYW\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "def display_file_head(file_path):\n",
    "    try:\n",
    "        file_name = os.path.basename(file_path)  \n",
    "        print(f\"\\n{file_name}:\\n\")  \n",
    "        with open(file_path, 'r') as file:\n",
    "            for _ in range(6): \n",
    "                print(file.readline().strip())\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {file_path}\")\n",
    "    except Exception as e:\n",
    "        print(f\"Error reading file {file_path}: {e}\")\n",
    "\n",
    "encode_metadata_file = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/filtered_metadata.txt\"\n",
    "encode_experiment_ids_file = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/ENCODE_experimentIDs.txt\"\n",
    "factorbook_motifs_metadata_file = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/factorbook_chipseq_meme_motifs.tsv\"\n",
    "mapping_output = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/bedtobigBed_mapping.txt\"\n",
    "\n",
    "print(\"### Input Files ###\")\n",
    "display_file_head(encode_metadata_file)\n",
    "display_file_head(factorbook_motifs_metadata_file)\n",
    "display_file_head(encode_experiment_ids_file)\n",
    "\n",
    "print(\"\\n### Output File ###\")\n",
    "display_file_head(mapping_output)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Download Factorbook motifs\n",
    "\n",
    "This script downloads factorbook motifs in BED format, by constructing download URLs using motif IDs from section 3.2\n",
    "\n",
    "Script: step3-download_facrotbook_motifs.py\n",
    "\n",
    "Input: bedtobigBed_mapping.txt from section 3.2\n",
    "\n",
    "Output: binding sites for the motif ENCFF675PHY_STGATGCAABC in BED format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Input Files ###\n",
      "\n",
      "bedtobigBed_mapping.txt:\n",
      "\n",
      "Animal\tBiosample\tTarget\tExperiment_ID\tExperiment_ID\tbigBed_Accession\tBED_Accession\tConsensus\tSample_Type\tBED_Consensus\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tAAAGAGGAASTGAAA\tcell line\tENCFF002XEC_AAAGAGGAASTGAAA\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tCCTCCTAGCCCTRACACACAGCTGGG\tcell line\tENCFF002XEC_CCTCCTAGCCCTRACACACAGCTGGG\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tSCBSBCMSBBCCRVCCHSCSVKCCYKSVCC\tcell line\tENCFF002XEC_SCBSBCMSBBCCRVCCHSCSVKCCYKSVCC\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tGCSYGGGGCAAGTGACCGTGCGTGTAAAGG\tcell line\tENCFF002XEC_GCSYGGGGCAAGTGACCGTGCGTGTAAAGG\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tTGGACTTTGRACYYW\tcell line\tENCFF002XEC_TGGACTTTGRACYYW\n",
      "\n",
      "### Output File ###\n",
      "\n",
      "ENCFF675PHY_STGATGCAABC.gz:\n",
      "\n",
      "chr11\t3057416\t3057427\t+\t0.0529\n",
      "chr6\t6655191\t6655202\t+\t0.0529\n",
      "chr2\t218745386\t218745397\t+\t0.0529\n",
      "chr19\t33374069\t33374080\t+\t0.0529\n",
      "chr1\t220046302\t220046313\t+\t0.0529\n",
      "chr10\t72272976\t72272987\t+\t0.0529\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import gzip\n",
    "\n",
    "def display_file_head(file_path):\n",
    "    try:\n",
    "        file_name = os.path.basename(file_path)  \n",
    "        print(f\"\\n{file_name}:\\n\")  \n",
    "        \n",
    "        if file_path.endswith('.gz'):\n",
    "            with gzip.open(file_path, 'rt') as file:\n",
    "                for _ in range(6):\n",
    "                    print(file.readline().strip())\n",
    "        else:\n",
    "            with open(file_path, 'r') as file:\n",
    "                for _ in range(6): \n",
    "                    print(file.readline().strip())\n",
    "                    \n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {file_path}\")\n",
    "    except Exception as e:\n",
    "        print(f\"Error reading file {file_path}: {e}\")\n",
    "\n",
    "mapping_file = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/bedtobigBed_mapping.txt\"\n",
    "downloaded_BED_file = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/fb_motifs/ENCFF675PHY_STGATGCAABC.gz\"\n",
    "\n",
    "print(\"### Input Files ###\")\n",
    "display_file_head(mapping_file)\n",
    "\n",
    "print(\"\\n### Output File ###\")\n",
    "display_file_head(downloaded_BED_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.4 Run Tomtom to compare factorbook motifs to HOCOMOCO and JASPAR known motifs and process the results\n",
    "\n",
    "This script runs Tomtom on factorbook motifs against HOCOMOCO and JASPAR motif databases, then process the output to identify the best motif matches based on q-value and sequence overlap.\n",
    "\n",
    "Script: step4-run_tomtom.py\n",
    "\n",
    "Input: \n",
    "1. complete-factorbook-catalog.meme from section 1.1\n",
    "2. HOCOMOCO v12 Core collection - H12CORE_meme_format.meme \n",
    "3. HOCOMOCO v12 core metadata - H12CORE_motifs.tsv\n",
    "4. JASPAR 2022 Core collection - JASPAR2022_CORE_non-redundant_v2.meme \n",
    "\n",
    "Output: tomtom_mapping.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Input Files ###\n",
      "\n",
      "complete-factorbook-catalog.meme:\n",
      "\n",
      "MEME version 4.5\n",
      "ALPHABET= ACGT\n",
      "\n",
      "MOTIF ENCSR437GBJ_AAAGAGGAASTGAAA\n",
      "letter-probability matrix: alength= 4 w= 15 nsites= 132 E= 4.400e-97\n",
      "0.5758 0.0000 0.2045 0.2197\n",
      "0.4848 0.1288 0.2045 0.1818\n",
      "0.5833 0.0000 0.1591 0.2576\n",
      "\n",
      "H12CORE_meme_format.meme:\n",
      "\n",
      "MEME version 4\n",
      "\n",
      "ALPHABET= ACGT\n",
      "\n",
      "strands: + -\n",
      "\n",
      "Background letter frequencies\n",
      "A 0.25 C 0.25 G 0.25 T 0.25\n",
      "\n",
      "JASPAR2022_CORE_non-redundant_v2.meme:\n",
      "\n",
      "MEME version 4\n",
      "\n",
      "ALPHABET= ACGT\n",
      "\n",
      "strands: + -\n",
      "\n",
      "Background letter frequencies\n",
      "A 0.25 C 0.25 G 0.25 T 0.25\n",
      "\n",
      "H12CORE_motifs.tsv:\n",
      "\n",
      "Motif\tLOGO\tGene (human)\tGene (mouse)\tQuality\tMotif subtype\tQuality\tTF family\tUniProt ID (human)\tUniProt ID (mouse)\n",
      "AHR.H12CORE.0.P.B\t\tAHR\tAhr\tB\t0\tP\tPAS {1.2.5}\tAHR_HUMAN\tAHR_MOUSE\n",
      "AHRR.H12CORE.0.P.C\t\tAHRR\tAhrr\tC\t0\tP\tPAS {1.2.5}\tAHRR_HUMAN\tAHRR_MOUSE\n",
      "ALX1.H12CORE.0.SM.B\t\tALX1\tAlx1\tB\t0\tS+M\tPaired-related HD {3.1.3}\tALX1_HUMAN\tALX1_MOUSE\n",
      "ALX3.H12CORE.0.SM.B\t\tALX3\tAlx3\tB\t0\tS+M\tPaired-related HD {3.1.3}\tALX3_HUMAN\tALX3_MOUSE\n",
      "ALX3.H12CORE.1.S.B\t\tALX3\tAlx3\tB\t1\tS\tPaired-related HD {3.1.3}\tALX3_HUMAN\tALX3_MOUSE\n",
      "ALX4.H12CORE.0.S.B\t\tALX4\tAlx4\tB\t0\tS\tPaired-related HD {3.1.3}\tALX4_HUMAN\tALX4_MOUSE\n",
      "ALX4.H12CORE.1.SM.B\t\tALX4\tAlx4\tB\t1\tS+M\tPaired-related HD {3.1.3}\tALX4_HUMAN\tALX4_MOUSE\n",
      "\n",
      "### Output File ###\n",
      "\n",
      "tomtom_mapping.txt:\n",
      "\n",
      "query_ID\tHOCOMOCO_ID\tHOCOMOCO_q-value\tHOCOMOCO_query_consensus\tHOCOMOCO_Target_consensus\tHOCOMOCO_TF_name\tHOCOMOCO_species\tJASPAR_ID\tJASPAR_q-value\tJASPAR_query_consensus\tJASPAR_Target_consensus\tJASPAR_TF_name\tJASPAR_species\n",
      "ENCSR000FAL_GCYGCYGCYGCCKCC\tZN519.H12CORE.0.P.C\t2.35385e-07\tGCCGCCGCCGCCGCC\tGCCGCCGCCGCCGCCGCCCC\tZN519\tZN519_HUMAN\tMA2022.1\t1.06006e-07\tGCCGCCGCCGCCGCC\tCCGCCGCCGCCGCCGCC\tLOB\tArabidopsis thaliana\n",
      "ENCSR000BUU_TTCAGCACCAYGGACA\tREST.H12CORE.0.P.B\t4.04185e-11\tTTCAGCACCATGGACA\tTTCAGCACCATGGACAGCGCCC\tREST\tREST_HUMAN\tMA0138.2\t2.80469e-11\tTTCAGCACCATGGACA\tTTCAGCACCATGGACAGCGCC\tREST\tHomo sapiens\n",
      "ENCSR796VBR_TGAGTCAY\tFOSB.H12CORE.0.P.B\t7.94831e-05\tTGAGTCAT\tATGAGTCAT\tFOSB\tFOSB_HUMAN\tMA1988.1\t0.000972578\tTGAGTCAT\tTATGAGTCATC\tAtf3\tMus musculus\n",
      "ENCSR534VHI_RRACTACAWYTCCCAGMAKGC\tno_match\tNA\tNA\tNA\tNA\tNA\tMA1573.2\t2.05455e-08\tGAACTACAATTCCCAGAATGC\tACTACAATTCCCAG\tThap11\tMus musculus\n",
      "ENCSR000BVA_BATKNCCAWATAWGGMAANN\tSRF.H12CORE.0.PSM.A\t3.17399e-09\tTATGCCCATATAAGGCAAAT\tATGCCCATATAAGGCAA\tSRF\tSRF_HUMAN\tno_match\tNA\tNA\tNA\tNA\tNA\n",
      "ENCSR000BKE_GTCACGTG\tTFE3.H12CORE.0.PSM.A\t5.0608e-05\tGTCACGTG\tGGTCACGTGAT\tTFE3\tTFE3_HUMAN\tMA0871.2\t0.00134658\tGTCACGTG\tGGTCACGTGGG\tTFEC\tHomo sapiens\n",
      "ENCSR000BTE_RTGACTCA\tATF3.H12CORE.0.P.B\t3.51441e-05\tATGACTCA\tGGATGACTCA\tATF3\tATF3_HUMAN\tMA0099.3\t1.65038e-05\tATGACTCA\tGATGACTCAT\tFOS::JUN\tHomo sapiens\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "def display_file_head(file_path):\n",
    "    try:\n",
    "        file_name = os.path.basename(file_path)  \n",
    "        print(f\"\\n{file_name}:\\n\")  \n",
    "        with open(file_path, 'r') as file:\n",
    "            for _ in range(8): \n",
    "                print(file.readline().strip())\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {file_path}\")\n",
    "    except Exception as e:\n",
    "        print(f\"Error reading file {file_path}: {e}\")\n",
    "\n",
    "fb_meme = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/complete-factorbook-catalog.meme\"\n",
    "hoco_meme = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/db/H12CORE_meme_format.meme\"\n",
    "jaspar_meme = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/db/JASPAR2022_CORE_non-redundant_v2.meme\"\n",
    "hoco_metadata = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/H12CORE_motifs.tsv\"\n",
    "\n",
    "processed_results = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/tomtom_mapping.txt\"\n",
    "\n",
    "print(\"### Input Files ###\")\n",
    "display_file_head(fb_meme)\n",
    "display_file_head(hoco_meme)\n",
    "display_file_head(jaspar_meme)\n",
    "display_file_head(hoco_metadata)\n",
    "\n",
    "print(\"\\n### Output File ###\")\n",
    "display_file_head(processed_results)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.5 Process factorbook motifs - Annotation and filtering intervals by p-value.\n",
    "\n",
    "This script processes factorbook motifs by integrating metadata from ENCODE, HOCOMOCO, and JASPAR databases. It annotates factorbook motifs with transcription factor information and generates both p-value-filtered and unfiltered datasets based on FIMO p-values.\n",
    "\n",
    "script: step5-motif_processing.py\n",
    "\n",
    "Input:\n",
    "1. Factorbook motifs from section 3.3\n",
    "2. bedtobigBed_mapping.txt from section 3.2\n",
    "3. tomtom_mapping.txt from section 3.4\n",
    "\n",
    "Output:\n",
    "1. fb_motifs_filtered - annotated intervals filtered for p-value < 1e-4.\n",
    "2. fb_motifs_processed - annotated intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Input Files ###\n",
      "\n",
      "ENCFF003ZLG_TYTCASTTCCTCWTT.gz:\n",
      "\n",
      "chr15\t77406322\t77406337\t-\t0.00161\n",
      "chr6\t43684212\t43684227\t-\t0.00161\n",
      "chr4\t134740490\t134740505\t-\t0.00161\n",
      "chr4\t88859044\t88859059\t-\t0.00161\n",
      "chr2\t172437883\t172437898\t-\t0.00161\n",
      "chr15\t77406322\t77406337\t-\t0.00161\n",
      "\n",
      "bedtobigBed_mapping.txt:\n",
      "\n",
      "Animal\tBiosample\tTarget\tExperiment_ID\tExperiment_ID\tbigBed_Accession\tBED_Accession\tConsensus\tSample_Type\tBED_Consensus\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tAAAGAGGAASTGAAA\tcell line\tENCFF002XEC_AAAGAGGAASTGAAA\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tCCTCCTAGCCCTRACACACAGCTGGG\tcell line\tENCFF002XEC_CCTCCTAGCCCTRACACACAGCTGGG\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tSCBSBCMSBBCCRVCCHSCSVKCCYKSVCC\tcell line\tENCFF002XEC_SCBSBCMSBBCCRVCCHSCSVKCCYKSVCC\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tGCSYGGGGCAAGTGACCGTGCGTGTAAAGG\tcell line\tENCFF002XEC_GCSYGGGGCAAGTGACCGTGCGTGTAAAGG\n",
      "Homo sapiens\tGM12878\tNFATC3\tENCSR437GBJ\tENCSR437GBJ\tENCFF340KVJ\tENCFF002XEC\tTGGACTTTGRACYYW\tcell line\tENCFF002XEC_TGGACTTTGRACYYW\n",
      "\n",
      "tomtom_mapping.txt:\n",
      "\n",
      "query_ID\tHOCOMOCO_ID\tHOCOMOCO_q-value\tHOCOMOCO_query_consensus\tHOCOMOCO_Target_consensus\tHOCOMOCO_TF_name\tHOCOMOCO_species\tJASPAR_ID\tJASPAR_q-value\tJASPAR_query_consensus\tJASPAR_Target_consensus\tJASPAR_TF_name\tJASPAR_species\n",
      "ENCSR000FAL_GCYGCYGCYGCCKCC\tZN519.H12CORE.0.P.C\t2.35385e-07\tGCCGCCGCCGCCGCC\tGCCGCCGCCGCCGCCGCCCC\tZN519\tZN519_HUMAN\tMA2022.1\t1.06006e-07\tGCCGCCGCCGCCGCC\tCCGCCGCCGCCGCCGCC\tLOB\tArabidopsis thaliana\n",
      "ENCSR000BUU_TTCAGCACCAYGGACA\tREST.H12CORE.0.P.B\t4.04185e-11\tTTCAGCACCATGGACA\tTTCAGCACCATGGACAGCGCCC\tREST\tREST_HUMAN\tMA0138.2\t2.80469e-11\tTTCAGCACCATGGACA\tTTCAGCACCATGGACAGCGCC\tREST\tHomo sapiens\n",
      "ENCSR796VBR_TGAGTCAY\tFOSB.H12CORE.0.P.B\t7.94831e-05\tTGAGTCAT\tATGAGTCAT\tFOSB\tFOSB_HUMAN\tMA1988.1\t0.000972578\tTGAGTCAT\tTATGAGTCATC\tAtf3\tMus musculus\n",
      "ENCSR534VHI_RRACTACAWYTCCCAGMAKGC\tno_match\tNA\tNA\tNA\tNA\tNA\tMA1573.2\t2.05455e-08\tGAACTACAATTCCCAGAATGC\tACTACAATTCCCAG\tThap11\tMus musculus\n",
      "ENCSR000BVA_BATKNCCAWATAWGGMAANN\tSRF.H12CORE.0.PSM.A\t3.17399e-09\tTATGCCCATATAAGGCAAAT\tATGCCCATATAAGGCAA\tSRF\tSRF_HUMAN\tno_match\tNA\tNA\tNA\tNA\tNA\n",
      "\n",
      "### Output File ###\n",
      "\n",
      "ENCFF003ZLG_TYTCASTTCCTCWTT_sorted.bed.gz:\n",
      "\n",
      "#chrom\tchromStart\tchromEnd\thocomoco_jaspar_TF_name\tFB_score\tstrand\tbinding_sequence\tFB_pwm_ID\tFB_consensus_sequence\tFB_ChIP-Seq_target\tHOCOMOCO_pwm_ID\tHOCOMOCO_Target_consensus\tHOCOMOCO_q-value\tJASPAR_pwm_ID\tJASPAR_Target_consensus\tJASPAR_q-value\tENCODE_exp_bigBed_bed_IDs\n",
      "chr1\t1013176\t1013191\tSPIB,SPIB\t0.138\t-\tTTTCATTTTCCATTT\tENCFF003ZLG_TYTCASTTCCTCWTT\tTYTCASTTCCTCWTT\tDPF2\tSPIB.H12CORE.0.P.B\tCTTTCACTTCCTCTTT\t4.45764e-10\tMA0081.2\tTTTCACTTCCTCTTTT\t1.09157e-08\tENCSR509FWH-ENCFF681AJV-ENCFF003ZLG\n",
      "chr1\t1013176\t1013191\tSPIB,SPIB\t0.138\t-\tTTTCATTTTCCATTT\tENCFF003ZLG_TYTCASTTCCTCWTT\tTYTCASTTCCTCWTT\tDPF2\tSPIB.H12CORE.0.P.B\tCTTTCACTTCCTCTTT\t4.45764e-10\tMA0081.2\tTTTCACTTCCTCTTTT\t1.09157e-08\tENCSR509FWH-ENCFF681AJV-ENCFF003ZLG\n",
      "chr1\t1013384\t1013399\tSPIB,SPIB\t0.121\t-\tTTTCGGTTTCCCTTT\tENCFF003ZLG_TYTCASTTCCTCWTT\tTYTCASTTCCTCWTT\tDPF2\tSPIB.H12CORE.0.P.B\tCTTTCACTTCCTCTTT\t4.45764e-10\tMA0081.2\tTTTCACTTCCTCTTTT\t1.09157e-08\tENCSR509FWH-ENCFF681AJV-ENCFF003ZLG\n",
      "chr1\t1013384\t1013399\tSPIB,SPIB\t0.121\t-\tTTTCGGTTTCCCTTT\tENCFF003ZLG_TYTCASTTCCTCWTT\tTYTCASTTCCTCWTT\tDPF2\tSPIB.H12CORE.0.P.B\tCTTTCACTTCCTCTTT\t4.45764e-10\tMA0081.2\tTTTCACTTCCTCTTTT\t1.09157e-08\tENCSR509FWH-ENCFF681AJV-ENCFF003ZLG\n",
      "chr1\t1013390\t1013405\tSPIB,SPIB\t0.161\t-\tCTTCAGTTTCGGTTT\tENCFF003ZLG_TYTCASTTCCTCWTT\tTYTCASTTCCTCWTT\tDPF2\tSPIB.H12CORE.0.P.B\tCTTTCACTTCCTCTTT\t4.45764e-10\tMA0081.2\tTTTCACTTCCTCTTTT\t1.09157e-08\tENCSR509FWH-ENCFF681AJV-ENCFF003ZLG\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import gzip\n",
    "\n",
    "def display_file_head(file_path):\n",
    "    try:\n",
    "        file_name = os.path.basename(file_path)  \n",
    "        print(f\"\\n{file_name}:\\n\")  \n",
    "        \n",
    "        if file_path.endswith('.gz'):\n",
    "            with gzip.open(file_path, 'rt') as file:\n",
    "                for _ in range(6):\n",
    "                    print(file.readline().strip())\n",
    "        else:\n",
    "            with open(file_path, 'r') as file:\n",
    "                for _ in range(6): \n",
    "                    print(file.readline().strip())\n",
    "                    \n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {file_path}\")\n",
    "    except Exception as e:\n",
    "        print(f\"Error reading file {file_path}: {e}\")\n",
    "\n",
    "fb_motif = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/fb_motifs/ENCFF003ZLG_TYTCASTTCCTCWTT.gz\"\n",
    "encode_mapping_file = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/bedtobigBed_mapping.txt\"\n",
    "tomtom_mapping = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/tomtom_mapping.txt\"\n",
    "\n",
    "unfiltered_output = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/fb_motifs_processed/ENCFF003ZLG_TYTCASTTCCTCWTT_sorted.bed.gz\" \n",
    "\n",
    "print(\"### Input Files ###\")\n",
    "display_file_head(fb_motif)\n",
    "display_file_head(encode_mapping_file)\n",
    "display_file_head(tomtom_mapping)\n",
    "\n",
    "\n",
    "print(\"\\n### Output File ###\")\n",
    "display_file_head(unfiltered_output)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### BED file columns\n",
      "\n",
      "#COLUMN\tDESCRIPTION\n",
      "\n",
      "chrom\tChromosome in chrXXX notation (e.g., chr1, chr2)\n",
      "chromStart\tStart position of the binding site. 0-based coordinate system\n",
      "chromEnd\tEnd position of the binding site.\n",
      "hocomoco_jaspar_TF_name\tTranscription factor names from HOCOMOCO and JASPAR databases. If both are available, they are comma-separated\n",
      "FB_score\tFIMO p-value\n",
      "strand\tStrand (+ for forward, - for reverse)\n",
      "binding_sequence\tStrand specific nucleotide sequence that binds to the transcription factor, extracted from the reference genome using the binding site coordinates (reverse complemented for '-' starnd)\n",
      "FB_pwm_ID\tFactorbook motif Position Weight Matrix (PWM) ID\n",
      "FB_consensus_sequence\tConsensus sequence representing the most common nucleotides at each position within the Factorbook motif\n",
      "FB_ChIP-Seq_target\tTarget of the ChIP-Seq experiment\n",
      "HOCOMOCO_pwm_ID\tHOCOMOCO motif Position Weight Matrix (PWM) ID\n",
      "HOCOMOCO_Target_consensus\tConsensus sequence from the HOCOMOCO database representing the target site of the transcription factor\n",
      "HOCOMOCO_q-value\tQ-value indicating the significance of the match between Factorbook motif and known motifs in the HOCOMOCO database.\n",
      "JASPAR_pwm_ID\tJASPAR motif Position Weight Matrix (PWM) ID\n",
      "JASPAR_Target_consensus\tConsensus sequence from the JASPAR database representing the target site of the transcription factor\n",
      "JASPAR_q-value\tQ-value indicating the significance of the match between Factorbook motif and known motifs in the JASPAR database.\n",
      "ENCODE_exp_bigBed_bed_IDs\tENCODE experiment ID, bigBed Accession, and BED Accession (hyphen - seperated)\n"
     ]
    }
   ],
   "source": [
    "def display_file(filepath):\n",
    "    with open(filepath, 'r') as file:\n",
    "        for line in file:\n",
    "            print(line.strip())\n",
    "\n",
    "header_file = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/bed17_column_description.tsv\"\n",
    "\n",
    "print('### BED file columns\\n')\n",
    "\n",
    "display_file(header_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.6 Factorbook motif probability matrix\n",
    "\n",
    "This script processes the factorbook motifs MEME catalog and writes each motif to a separate MEME file.\n",
    "\n",
    "Script: step6-pwm-proecssing.py\n",
    "\n",
    "Input: complete-factorbook-catalog.meme.gz from section 1.1\n",
    "\n",
    "Output: ENCSR444LIN_CTTTGAAR.meme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### complete factorbook catalog\n",
      "\n",
      "complete-factorbook-catalog.meme:\n",
      "\n",
      "MEME version 4.5\n",
      "ALPHABET= ACGT\n",
      "\n",
      "MOTIF ENCSR437GBJ_AAAGAGGAASTGAAA\n",
      "letter-probability matrix: alength= 4 w= 15 nsites= 132 E= 4.400e-97\n",
      "0.5758 0.0000 0.2045 0.2197\n",
      "0.4848 0.1288 0.2045 0.1818\n",
      "0.5833 0.0000 0.1591 0.2576\n",
      "0.2500 0.0682 0.5758 0.1061\n",
      "0.6515 0.0758 0.1742 0.0985\n",
      "0.2348 0.0000 0.7652 0.0000\n",
      "0.2045 0.0076 0.7879 0.0000\n",
      "0.9167 0.0000 0.0682 0.0152\n",
      "0.9773 0.0076 0.0000 0.0152\n",
      "0.1742 0.3788 0.4470 0.0000\n",
      "0.2424 0.0152 0.0379 0.7045\n",
      "0.0758 0.0000 0.9242 0.0000\n",
      "0.9167 0.0000 0.0530 0.0303\n",
      "0.6742 0.0000 0.2803 0.0455\n",
      "0.7727 0.0530 0.1288 0.0455\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "def display_file_head(file_path):\n",
    "    try:\n",
    "        file_name = os.path.basename(file_path)  \n",
    "        print(f\"\\n{file_name}:\\n\")  \n",
    "        with open(file_path, 'r') as file:\n",
    "            for _ in range(20): \n",
    "                print(file.readline().strip())\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {file_path}\")\n",
    "    except Exception as e:\n",
    "        print(f\"Error reading file {file_path}: {e}\")\n",
    "\n",
    "fb_meme = \"/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/complete-factorbook-catalog.meme\"\n",
    "\n",
    "print('### complete factorbook catalog')\n",
    "\n",
    "display_file_head(fb_meme)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
