{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FACTORBOOK data processing SOP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Introduction\n",
    "\n",
    "Factorbook is a transcription factor (TF)-centric repository of all ENCODE ChIP-seq datasets on TF-binding regions, as well as the rich analysis results of these data.\n",
    "\n",
    "#### 1.1 Factorbook motifs:\n",
    "Motifs were identified by applying MEME to the top 500 IDR thresholded ChIP-seq peaks from more than 3000 ENCODE ChIP-seq experiments. Five motifs were identified per experiment. These motifs were then filtered for quality using peak centrality and enrichment metrics. In total, there are 6,069 motifs available.\n",
    "\n",
    "[Factorbook motif catalog](https://downloads.wenglab.org/factorbook-download/complete-factorbook-catalog.meme.gz)\n",
    "\n",
    "[Associated metadata](https://storage.googleapis.com/gcp.wenglab.org/factorbook_chipseq_meme_motifs.tsv)\n",
    "\n",
    "#### 1.2. Motif annotation:\n",
    "\n",
    "Factorbook motifs are downloaded and processed to include overlaps with known motifs from the HOCOMOCO and JASPAR databases.\n",
    "\n",
    "TOMTOM was used to compare Factorbook MEME motifs against the HOCOMOCO and JASPAR catalogs. Tomtom results were filtered for matches where the query is a part of the target or vice versa. From these filtered results, motifs with the lowest q-value were selected as the most confident overlaps between Factorbook motifs and the known motifs in HOCOMOCO and JASPAR.\n",
    "\n",
    "The purpose of this SOP is to document the stpes and processes involved in processing and annotation Facotbook motifs.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## 2.Requirements\n",
    "\n",
    "### 2.1 ENCODE metadata:\n",
    "\n",
    "ENCODE metadata is necessary to construct the download URL for each of the factorbook motifs. For example, factorbook motif ID (e.g., ENCFF674XTY_CMTBCTGGGARTTGTAG) is generated by combining the ENCODE BED file ID with the consensus sequence. \n",
    "\n",
    "#### 2.1.1 metadata download:\n",
    "Select all the bed and bigBed tracks available in the ENCODE experiment matrix, using the below selection criteria:\n",
    "\n",
    "- Biosample: Organism --> Homo sapiens\n",
    "- Quality: Status --> released, archived, revoked\n",
    "- Analysis: Available file types --> All bed anf bigBed file types\n",
    "- Analysis: Genome Assembly --> hg19, GRCh38\n",
    "\n",
    "The following URL is pre-selected to meet the above criteria on the ENCODE data portal:\n",
    "\n",
    "[ENCODE metadata](https://www.encodeproject.org/search/?type=Experiment&status=released&perturbed=false&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens&perturbed=true&status=archived&status=revoked&files.file_type=bed+narrowPeak&files.file_type=bigBed+narrowPeak&files.file_type=bed+idr_ranked_peak&files.file_type=bed+broadPeak&files.file_type=bed+bed3%2B&files.file_type=bigBed+broadPeak&files.file_type=bigBed+bed3%2B&files.file_type=bed+bedRnaElements&files.file_type=bigBed+bedRnaElements&files.file_type=bed+tss_peak&files.file_type=bigBed+tss_peak&files.file_type=bed+bedMethyl&files.file_type=bigBed+bedMethyl&files.file_type=bed+bed9%2B&files.file_type=bed+bedGraph&files.file_type=bigInteract&files.file_type=bed+idr_peak&files.file_type=bigBed+idr_peak&files.file_type=bigBed+bed9%2B&files.file_type=bed+bed9&files.file_type=bigBed+bed9&files.file_type=bed+bedLogR&files.file_type=bigBed+bedLogR&files.file_type=bed+bed12&files.file_type=bigBed+bed12&files.file_type=bigBed+bed6%2B&files.file_type=bed+bed6%2B&files.file_type=bed+bedExonScore&files.file_type=bigBed+bedExonScore&files.file_type=bed+peptideMapping&files.file_type=bigBed+peptideMapping&files.file_type=bed+bed3&files.file_type=bigBed+bed3&files.file_type=bed+modPepMap&files.file_type=bed+pepMap&files.file_type=bigBed+modPepMap&files.file_type=bigBed+pepMap)\n",
    "\n",
    "### 2.2 Tomtom: motif comparison tool\n",
    "\n",
    "Tomtom compares one or more motifs against a database of known motifs (e.g., JASPAR). \n",
    "\n",
    "[MEME suite V5.5.5](https://meme-suite.org/meme/doc/download.html)\n",
    "\n",
    "### 2.3 Samtools\n",
    "\n",
    "[Samtools V1.17](https://www.htslib.org/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "## 3. FILER data processing pipeline\n",
    "\n",
    "### 3.1 Collect ENCODE experiment IDs from Factorbook motif catalog metadata\n",
    "\n",
    "Script: step1-get_ENCODE_experiments.sh\n",
    "\n",
    "Input: factorbook motif metadata - factorbook_chipseq_meme_motifs.tsv"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "\n",
    "Animal  Biosample       Target  Identifier      Consensus       Sample Type     Information\n",
    "Homo sapiens    GM12878 NFATC3  ENCSR437GBJ     AAAGAGGAASTGAAA cell line       \"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
    "Homo sapiens    GM12878 NFATC3  ENCSR437GBJ     CCTCCTAGCCCTRACACACAGCTGGG      cell line       \"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
    "Homo sapiens    GM12878 NFATC3  ENCSR437GBJ     SCBSBCMSBBCCRVCCHSCSVKCCYKSVCC  cell line       \"inconsistent platforms,mild to moderate bottlenecking,moderate library complexity,mild to moderate bottlenecking,mild to moderate bottlenecking\"\n",
    "Homo sapiens    HepG2   TIGD3   ENCSR725QZQ     AWANAAWHAMHTTWMWGACHAGAAAKRTTR  cell line       \"mild to moderate bottlenecking,mild to moderate bottlenecking,mild to moderate bottlenecking,mild to moderate bottlenecking,borderline replicate concordance,borderline replicate concordance,missing genetic modification reagents\"\n",
    "Homo sapiens    HepG2   TIGD3   ENCSR725QZQ     WGGTGCTGAAA     cell line       \"mild to moderate bottlenecking,mild to moderate bottlenecking,mild to moderate bottlenecking,mild to moderate bottlenecking,borderline replicate concordance,borderline replicate concordance,missing genetic modification reagents\"\n",
    "Homo sapiens    HepG2   MED13   ENCSR269DQN     RAAGATCAAAG     cell line       missing genetic modification reagents\n",
    "Homo sapiens    HepG2   MED13   ENCSR269DQN     CAGCACCRYGGACAGCGCC     cell line       missing genetic modification reagents\n",
    "Homo sapiens    HepG2   MED13   ENCSR269DQN     CCCCBCCCBCNCCYBBSCCCC   cell line       missing genetic modification reagents\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output: ENCODE_experimentIDs.txt"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "ENCODE_experiment\n",
    "ENCSR000AHD\n",
    "ENCSR000AHF\n",
    "ENCSR000AIB\n",
    "ENCSR000AIC\n",
    "ENCSR000AID\n",
    "ENCSR000AIE\n",
    "ENCSR000AIG\n",
    "ENCSR000AIH"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 process ENCODE metadata to map IDR thresholded peak BED file IDs for all ChIP-seq experiments\n",
    "\n",
    "Script: step2-bedtobigBed_mapping.py\n",
    "\n",
    "Input: \n",
    "1. ENCODE metadata\n",
    "2. ENCODE experiment IDs from step 3.1\n",
    "3. factorbook motifs metadata\n",
    "\n",
    "Output: bedtobigBed_mapping.txt"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "Animal  Biosample       Target  Experiment_ID   Experiment_ID   bigBed_Accession        BED_Accession   Consensus       Sample_Type     BED_Consensus\n",
    "Homo sapiens    GM12878 NFATC3  ENCSR437GBJ     ENCSR437GBJ     ENCFF340KVJ     ENCFF002XEC     AAAGAGGAASTGAAA cell line       ENCFF002XEC_AAAGAGGAASTGAAA\n",
    "Homo sapiens    GM12878 NFATC3  ENCSR437GBJ     ENCSR437GBJ     ENCFF340KVJ     ENCFF002XEC     CCTCCTAGCCCTRACACACAGCTGGG      cell line       ENCFF002XEC_CCTCCTAGCCCTRACACACAGCTGGG\n",
    "Homo sapiens    GM12878 NFATC3  ENCSR437GBJ     ENCSR437GBJ     ENCFF340KVJ     ENCFF002XEC     SCBSBCMSBBCCRVCCHSCSVKCCYKSVCC  cell line       ENCFF002XEC_SCBSBCMSBBCCRVCCHSCSVKCCYKSVCC\n",
    "Homo sapiens    GM12878 NFATC3  ENCSR437GBJ     ENCSR437GBJ     ENCFF340KVJ     ENCFF002XEC     GCSYGGGGCAAGTGACCGTGCGTGTAAAGG  cell line       ENCFF002XEC_GCSYGGGGCAAGTGACCGTGCGTGTAAAGG\n",
    "Homo sapiens    HepG2   TIGD3   ENCSR725QZQ     ENCSR725QZQ     ENCFF491KVL     ENCFF003AMQ     WGGTGCTGAAA     cell line       ENCFF003AMQ_WGGTGCTGAAA\n",
    "Homo sapiens    HepG2   TIGD3   ENCSR725QZQ     ENCSR725QZQ     ENCFF491KVL     ENCFF003AMQ     WTCAGCACCANGGACAGCDCC   cell line       ENCFF003AMQ_WTCAGCACCANGGACAGCDCC\n",
    "Homo sapiens    HepG2   TIGD3   ENCSR725QZQ     ENCSR725QZQ     ENCFF491KVL     ENCFF003AMQ     CCTYYYYCYBCCTBCTCTYYC   cell line       ENCFF003AMQ_CCTYYYYCYBCCTBCTCTYYC"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
