{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FACTORBOOK data processing SOP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Introduction\n",
    "\n",
    "Factorbook is a transcription factor (TF)-centric repository of all ENCODE ChIP-seq datasets on TF-binding regions, as well as the rich analysis results of these data.\n",
    "\n",
    "#### 1.1 Factorbook motifs:\n",
    "Motifs were identified by applying MEME to the top 500 IDR thresholded ChIP-seq peaks from more than 3000 ENCODE ChIP-seq experiments. Five motifs were identified per experiment. These motifs were then filtered for quality using peak centrality and enrichment metrics. In total, there are 6,069 motifs available.\n",
    "\n",
    "[Factorbook motif catalog](https://downloads.wenglab.org/factorbook-download/complete-factorbook-catalog.meme.gz)\n",
    "\n",
    "[Associated metadata](https://storage.googleapis.com/gcp.wenglab.org/factorbook_chipseq_meme_motifs.tsv)\n",
    "\n",
    "#### 1.2. Factorbook motif annotation for FILER\n",
    "\n",
    "Factorbook motifs are downloaded and processed to include overlaps with known motifs from the HOCOMOCO and JASPAR databases.\n",
    "\n",
    "TOMTOM was used to compare Factorbook MEME motifs against the HOCOMOCO and JASPAR catalogs. Tomtom results were filtered for matches where the query is a part of the target or vice versa. From these filtered results, motifs with the lowest q-value were selected as the most confident overlaps between Factorbook motifs and the known motifs in HOCOMOCO and JASPAR.\n",
    "\n",
    "The purpose of this SOP is to document the steps and processes involved in processing and annotating Facorbook motifs.\n",
    "\n",
    "#### 1.3 Data processing workflow\n",
    "\n",
    "![title](/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/FB.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## 2.Requirements\n",
    "\n",
    "### 2.1 ENCODE metadata:\n",
    "\n",
    "ENCODE metadata is necessary to construct the download URL for each of the factorbook motifs. For example, factorbook motif ID (e.g., ENCFF674XTY_CMTBCTGGGARTTGTAG) is generated by combining the ENCODE BED file ID with the consensus sequence. \n",
    "\n",
    "#### 2.1.1 metadata download:\n",
    "Select all the bed and bigBed tracks available in the ENCODE experiment matrix. Used the below selection criteria:\n",
    "\n",
    "- Biosample: Organism --> Homo sapiens\n",
    "- Quality: Status --> released, archived, revoked\n",
    "- Analysis: Available file types --> All bed and bigBed file types\n",
    "- Analysis: Genome Assembly --> hg19, GRCh38\n",
    "\n",
    "The following URL is pre-selected to meet the above criteria on the ENCODE data portal:\n",
    "\n",
    "[ENCODE metadata](https://www.encodeproject.org/search/?type=Experiment&status=released&perturbed=false&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens&perturbed=true&status=archived&status=revoked&files.file_type=bed+narrowPeak&files.file_type=bigBed+narrowPeak&files.file_type=bed+idr_ranked_peak&files.file_type=bed+broadPeak&files.file_type=bed+bed3%2B&files.file_type=bigBed+broadPeak&files.file_type=bigBed+bed3%2B&files.file_type=bed+bedRnaElements&files.file_type=bigBed+bedRnaElements&files.file_type=bed+tss_peak&files.file_type=bigBed+tss_peak&files.file_type=bed+bedMethyl&files.file_type=bigBed+bedMethyl&files.file_type=bed+bed9%2B&files.file_type=bed+bedGraph&files.file_type=bigInteract&files.file_type=bed+idr_peak&files.file_type=bigBed+idr_peak&files.file_type=bigBed+bed9%2B&files.file_type=bed+bed9&files.file_type=bigBed+bed9&files.file_type=bed+bedLogR&files.file_type=bigBed+bedLogR&files.file_type=bed+bed12&files.file_type=bigBed+bed12&files.file_type=bigBed+bed6%2B&files.file_type=bed+bed6%2B&files.file_type=bed+bedExonScore&files.file_type=bigBed+bedExonScore&files.file_type=bed+peptideMapping&files.file_type=bigBed+peptideMapping&files.file_type=bed+bed3&files.file_type=bigBed+bed3&files.file_type=bed+modPepMap&files.file_type=bed+pepMap&files.file_type=bigBed+modPepMap&files.file_type=bigBed+pepMap)\n",
    "\n",
    "### 2.2 Tomtom: Motif comparison tool\n",
    "\n",
    "Tomtom compares one or more motifs against a database of known motifs (e.g., JASPAR). \n",
    "\n",
    "[MEME suite V5.5.5](https://meme-suite.org/meme/doc/download.html)\n",
    "\n",
    "### 2.3 Samtools\n",
    "\n",
    "[Samtools V1.17](https://www.htslib.org/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. FILER data processing pipeline\n",
    "\n",
    "![title](/mnt/ebs/jackal/FILER2/FILER2-production/FACTORBOOK/fb_pipeline.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "### 3.1 Collect ENCODE experiment IDs from Factorbook motif catalog metadata\n",
    "\n",
    "Script: step1-get_ENCODE_experiments.sh\n",
    "\n",
    "Input: factorbook motif metadata - factorbook_chipseq_meme_motifs.tsv\n",
    "\n",
    "| Animal       | Biosample | Target | Identifier     | Consensus                 | Sample Type | Information                                                                                                                                                  |\n",
    "|--------------|-----------|--------|----------------|---------------------------|-------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "| Homo sapiens | GM12878   | NFATC3 | ENCSR437GBJ     | AAAGAGGAASTGAAA            | cell line   | \"inconsistent platforms, mild to moderate bottlenecking, moderate library complexity, mild to moderate bottlenecking, mild to moderate bottlenecking\"          |\n",
    "| Homo sapiens | GM12878   | NFATC3 | ENCSR437GBJ     | CCTCCTAGCCCTRACACACAGCTGGG | cell line   | \"inconsistent platforms, mild to moderate bottlenecking, moderate library complexity, mild to moderate bottlenecking, mild to moderate bottlenecking\"          |\n",
    "| Homo sapiens | HepG2     | TIGD3  | ENCSR725QZQ     | AWANAAWHAMHTTWMWGACHAGAAAKRTTR  | cell line   | \"mild to moderate bottlenecking, mild to moderate bottlenecking, mild to moderate bottlenecking, mild to moderate bottlenecking, borderline replicate concordance, missing genetic modification reagents\" |\n",
    "| Homo sapiens | HepG2     | TIGD3  | ENCSR725QZQ     | WGGTGCTGAAA                | cell line   | \"mild to moderate bottlenecking, mild to moderate bottlenecking, mild to moderate bottlenecking, mild to moderate bottlenecking, borderline replicate concordance, missing genetic modification reagents\" |\n",
    "| Homo sapiens | HepG2     | MED13  | ENCSR269DQN     | RAAGATCAAAG                | cell line   | missing genetic modification reagents                                                                                                                         |\n",
    "| Homo sapiens | HepG2     | MED13  | ENCSR269DQN     | CAGCACCRYGGACAGCGCC        | cell line   | missing genetic modification reagents                                                                                                                         |\n",
    "| Homo sapiens | HepG2     | MED13  | ENCSR269DQN     | CCCCBCCCBCNCCYBBSCCCC      | cell line   | missing genetic modification reagents                                                                                                                         |\n",
    "\n",
    "\n",
    "Output: ENCODE_experimentIDs.txt\n",
    "\n",
    "- ENCSR000AHD\n",
    "- ENCSR000AHF\n",
    "- ENCSR000AIB\n",
    "- ENCSR000AIC\n",
    "- ENCSR000AID\n",
    "- ENCSR000AIE\n",
    "- ENCSR000AIG\n",
    "- ENCSR000AIH"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Process ENCODE metadata to map IDR thresholded peak BED file IDs for all ChIP-seq experiments\n",
    "\n",
    "Script: step2-bedtobigBed_mapping.py\n",
    "\n",
    "Input: \n",
    "1. ENCODE metadata from section 2.1\n",
    "2. ENCODE experiment IDs from section 3.1\n",
    "3. factorbook motifs metadata section 1.1\n",
    "\n",
    "Output: bedtobigBed_mapping.txt\n",
    "\n",
    "| Animal       | Biosample | Target | Experiment_ID | Experiment_ID | bigBed_Accession | BED_Accession | Consensus                        | Sample_Type | BED_Consensus                        |\n",
    "|--------------|-----------|--------|---------------|---------------|------------------|---------------|-----------------------------------|-------------|--------------------------------------|\n",
    "| Homo sapiens | GM12878   | NFATC3 | ENCSR437GBJ   | ENCSR437GBJ   | ENCFF340KVJ      | ENCFF002XEC   | AAAGAGGAASTGAAA                  | cell line   | ENCFF002XEC_AAAGAGGAASTGAAA         |\n",
    "| Homo sapiens | GM12878   | NFATC3 | ENCSR437GBJ   | ENCSR437GBJ   | ENCFF340KVJ      | ENCFF002XEC   | CCTCCTAGCCCTRACACACAGCTGGG        | cell line   | ENCFF002XEC_CCTCCTAGCCCTRACACACAGCTGGG |\n",
    "| Homo sapiens | GM12878   | NFATC3 | ENCSR437GBJ   | ENCSR437GBJ   | ENCFF340KVJ      | ENCFF002XEC   | SCBSBCMSBBCCRVCCHSCSVKCCYKSVCC    | cell line   | ENCFF002XEC_SCBSBCMSBBCCRVCCHSCSVKCCYKSVCC |\n",
    "| Homo sapiens | GM12878   | NFATC3 | ENCSR437GBJ   | ENCSR437GBJ   | ENCFF340KVJ      | ENCFF002XEC   | GCSYGGGGCAAGTGACCGTGCGTGTAAAGG    | cell line   | ENCFF002XEC_GCSYGGGGCAAGTGACCGTGCGTGTAAAGG |\n",
    "| Homo sapiens | HepG2     | TIGD3  | ENCSR725QZQ   | ENCSR725QZQ   | ENCFF491KVL      | ENCFF003AMQ   | WGGTGCTGAAA                      | cell line   | ENCFF003AMQ_WGGTGCTGAAA             |\n",
    "| Homo sapiens | HepG2     | TIGD3  | ENCSR725QZQ   | ENCSR725QZQ   | ENCFF491KVL      | ENCFF003AMQ   | WTCAGCACCANGGACAGCDCC            | cell line   | ENCFF003AMQ_WTCAGCACCANGGACAGCDCC   |\n",
    "| Homo sapiens | HepG2     | TIGD3  | ENCSR725QZQ   | ENCSR725QZQ   | ENCFF491KVL      | ENCFF003AMQ   | CCTYYYYCYBCCTBCTCTYYC            | cell line   | ENCFF003AMQ_CCTYYYYCYBCCTBCTCTYYC   |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Download Factorbook motifs\n",
    "\n",
    "Script: step3-download_facrotbook_motifs.py\n",
    "\n",
    "Input: bedtobigBed_mapping.txt from section 3.2\n",
    "\n",
    "Output: binding sites for the motif ENCFF675PHY_STGATGCAABC in BED format\n",
    "\n",
    "| Chromosome | Start      | End        | Strand | p-value  |\n",
    "|------------|------------|------------|--------|--------|\n",
    "| chr11      | 3057416    | 3057427    | +      | 0.0529 |\n",
    "| chr6       | 6655191    | 6655202    | +      | 0.0529 |\n",
    "| chr2       | 218745386  | 218745397  | +      | 0.0529 |\n",
    "| chr19      | 33374069   | 33374080   | +      | 0.0529 |\n",
    "| chr1       | 220046302  | 220046313  | +      | 0.0529 |\n",
    "| chr10      | 72272976   | 72272987   | +      | 0.0529 |\n",
    "| chr17      | 41689493   | 41689504   | -      | 0.0529 |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.4 Run Tomtom to compare factorbook motifs to HOCOMOCO and JASPAR known motifs and process the results\n",
    "\n",
    "Script: step4-run_tomtom.py\n",
    "\n",
    "Input: \n",
    "1. complete-factorbook-catalog.meme from section 1.1\n",
    "2. HOCOMOCO v12 Core collection - H12CORE_meme_format.meme \n",
    "3. HOCOMOCO v12 core metadata - H12CORE_motifs.tsv\n",
    "4. JASPAR 2022 Core collection - JASPAR2022_CORE_non-redundant_v2.meme \n",
    "\n",
    "Output: tomtom_mapping.txt\n",
    "\n",
    "| query_ID                     | HOCOMOCO_ID           | HOCOMOCO_q-value | HOCOMOCO_query_consensus | HOCOMOCO_Target_consensus | HOCOMOCO_TF_name | HOCOMOCO_species | JASPAR_ID   | JASPAR_q-value | JASPAR_query_consensus | JASPAR_Target_consensus | JASPAR_TF_name | JASPAR_species        |\n",
    "|------------------------------|-----------------------|------------------|--------------------------|---------------------------|------------------|------------------|-------------|----------------|------------------------|-------------------------|----------------|-----------------------|\n",
    "| ENCSR000FAL_GCYGCYGCYGCCKCC   | ZN519.H12CORE.0.P.C   | 2.35385e-07      | GCCGCCGCCGCCGCC           | GCCGCCGCCGCCGCCGCCCC       | ZN519            | ZN519_HUMAN      | MA2022.1    | 1.06006e-07    | GCCGCCGCCGCCGCC        | CCGCCGCCGCCGCCGCC        | LOB            | Arabidopsis thaliana   |\n",
    "| ENCSR000BUU_TTCAGCACCAYGGACA  | REST.H12CORE.0.P.B    | 4.04185e-11      | TTCAGCACCATGGACA          | TTCAGCACCATGGACAGCGCCC     | REST             | REST_HUMAN       | MA0138.2    | 2.80469e-11    | TTCAGCACCATGGACA        | TTCAGCACCATGGACAGCGCC    | REST           | Homo sapiens          |\n",
    "| ENCSR796VBR_TGAGTCAY          | FOSB.H12CORE.0.P.B    | 7.94831e-05      | TGAGTCAT                  | ATGAGTCAT                  | FOSB             | FOSB_HUMAN       | MA1988.1    | 0.000972578    | TGAGTCAT                | TATGAGTCATC             | Atf3           | Mus musculus          |\n",
    "| ENCSR000BKE_GTCACGTG          | TFE3.H12CORE.0.PSM.A  | 5.0608e-05       | GTCACGTG                  | GGTCACGTGAT                | TFE3             | TFE3_HUMAN       | MA0871.2    | 0.00134658     | GTCACGTG                | GGTCACGTGGG             | TFEC           | Homo sapiens          |\n",
    "| ENCSR000BTE_RTGACTCA          | ATF3.H12CORE.0.P.B    | 3.51441e-05      | ATGACTCA                  | GGATGACTCA                 | ATF3             | ATF3_HUMAN       | MA0099.3    | 1.65038e-05    | ATGACTCA                | GATGACTCAT             | FOS::JUN       | Homo sapiens          |\n",
    "| ENCSR705HGT_YKCYSATTGGCYR     | no_match              | NA               | NA                        | NA                         | NA               | NA               | MA1644.1    | 6.0952e-06     | CTCTGATTGGCTG           | TCTGATTGGCT             | NFYC           | Homo sapiens          |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.5 Process factorbook motifs - Annotation and filtering intervals by p-value.\n",
    "\n",
    "script: step5-motif_processing.py\n",
    "\n",
    "Input:\n",
    "1. Factorbook motifs from section 3.3\n",
    "2. bedtobigBed_mapping.txt from section 3.2\n",
    "3. tomtom_mapping.txt from section 3.4\n",
    "\n",
    "Output:\n",
    "1. fb_motifs_filtered - annotated intervals filtered for p-value < 1e-4.\n",
    "2. fb_motifs_processed - annotated intervals\n",
    "\n",
    "| #chrom | chromStart | chromEnd | hocomoco_jaspar_TF_name | FB_score | strand | binding_sequence | FB_pwm_ID                | FB_consensus_sequence | FB_ChIP-Seq_target | HOCOMOCO_pwm_ID          | HOCOMOCO_Target_consensus | HOCOMOCO_q-value | JASPAR_pwm_ID | JASPAR_Target_consensus | JASPAR_q-value | ENCODE_exp_bigBed_bed_IDs                      |\n",
    "|--------|------------|----------|-------------------------|----------|--------|------------------|--------------------------|-----------------------|--------------------|--------------------------|--------------------------|-----------------|---------------|------------------------|----------------|------------------------------------------------|\n",
    "| chr1   | 115809     | 115817   | GATA1,GATA1::TAL1        | 0.178    | -      | CTTATCTG          | ENCFF321NXM_CTTATCWS      | CTTATCWS              | NCOR1              | GATA1.H12CORE.1.PSM.A     | CCTTATCTGC               | 4.6956e-05      | MA0140.2      | CTTATCTGTGAGGAGCAG      | 0.0155726      | ENCSR798ILC-ENCFF866HRM-ENCFF321NXM            |\n",
    "| chr1   | 1109904    | 1109912  | GATA1,GATA1::TAL1        | 0.185    | +      | CTTATCAG          | ENCFF321NXM_CTTATCWS      | CTTATCWS              | NCOR1              | GATA1.H12CORE.1.PSM.A     | CCTTATCTGC               | 4.6956e-05      | MA0140.2      | CTTATCTGTGAGGAGCAG      | 0.0155726      | ENCSR798ILC-ENCFF866HRM-ENCFF321NXM            |\n",
    "| chr1   | 1883636    | 1883644  | GATA1,GATA1::TAL1        | 0.178    | -      | CTTATCTG          | ENCFF321NXM_CTTATCWS      | CTTATCWS              | NCOR1              | GATA1.H12CORE.1.PSM.A     | CCTTATCTGC               | 4.6956e-05      | MA0140.2      | CTTATCTGTGAGGAGCAG      | 0.0155726      | ENCSR798ILC-ENCFF866HRM-ENCFF321NXM            |\n",
    "| chr1   | 2585948    | 2585956  | GATA1,GATA1::TAL1        | 0.185    | +      | CTTATCTC          | ENCFF321NXM_CTTATCWS      | CTTATCWS              | NCOR1              | GATA1.H12CORE.1.PSM.A     | CCTTATCTGC               | 4.6956e-05      | MA0140.2      | CTTATCTGTGAGGAGCAG      | 0.0155726      | ENCSR798ILC-ENCFF866HRM-ENCFF321NXM            |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| COLUMN                    | DESCRIPTION                                                                                                                                             |\n",
    "|---------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "| chrom                     | Chromosome in chrXXX notation (e.g., chr1, chr2)                                                                                                        |\n",
    "| chromStart                | Start position of the binding site. 0-based coordinate system                                                                                            |\n",
    "| chromEnd                  | End position of the binding site.                                                                                                                        |\n",
    "| hocomoco_jaspar_TF_name    | Transcription factor names from HOCOMOCO and JASPAR databases. If both are available, they are comma-separated                                           |\n",
    "| FB_score                  | FIMO p-value                                                                                                                                            |\n",
    "| strand                    | Strand (+ for forward, - for reverse)                                                                                                                    |\n",
    "| binding_sequence          | Strand specific nucleotide sequence that binds to the transcription factor, extracted from the reference genome using the binding site coordinates (reverse complemented for '-' strand) |\n",
    "| FB_pwm_ID                 | Factorbook motif Position Weight Matrix (PWM) ID                                                                                                         |\n",
    "| FB_consensus_sequence      | Consensus sequence representing the most common nucleotides at each position within the Factorbook motif                                                 |\n",
    "| FB_ChIP-Seq_target         | Target of the ChIP-Seq experiment                                                                                                                        |\n",
    "| HOCOMOCO_pwm_ID            | HOCOMOCO motif Position Weight Matrix (PWM) ID                                                                                                           |\n",
    "| HOCOMOCO_Target_consensus  | Consensus sequence from the HOCOMOCO database representing the target site of the transcription factor                                                    |\n",
    "| HOCOMOCO_q-value           | Q-value indicating the significance of the match between Factorbook motif and known motifs in the HOCOMOCO database.                                      |\n",
    "| JASPAR_pwm_ID              | JASPAR motif Position Weight Matrix (PWM) ID                                                                                                             |\n",
    "| JASPAR_Target_consensus    | Consensus sequence from the JASPAR database representing the target site of the transcription factor                                                      |\n",
    "| JASPAR_q-value             | Q-value indicating the significance of the match between Factorbook motif and known motifs in the JASPAR database.                                        |\n",
    "| ENCODE_exp_bigBed_bed_IDs  | ENCODE experiment ID, bigBed Accession, and BED Accession (hyphen-separated)                                                                             |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.6 Factorbook motif probability matrix\n",
    "\n",
    "Script: step6-pwm-proecssing.py\n",
    "\n",
    "Input: complete-factorbook-catalog.meme.gz from section 1.1\n",
    "\n",
    "Output: ENCSR444LIN_CTTTGAAR.meme"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "MEME version 4.5\n",
    "ALPHABET= ACGT\n",
    "\n",
    "MOTIF ENCSR444LIN_CTTTGAAR\n",
    "letter-probability matrix: alength= 4 w= 8 nsites= 33 E= 2.100e-02\n",
    "0.0000 1.0000 0.0000 0.0000\n",
    "0.0000 0.0000 0.0000 1.0000\n",
    "0.0000 0.0000 0.0000 1.0000\n",
    "0.0000 0.0000 0.0000 1.0000\n",
    "0.0000 0.0000 1.0000 0.0000\n",
    "1.0000 0.0000 0.0000 0.0000\n",
    "1.0000 0.0000 0.0000 0.0000\n",
    "0.4848 0.0000 0.5152 0.0000"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
